### DISCUSSION TITLE
Is there a way to run multiple statements/transaction/a sql script ?

### DISCUSSION THREAD

**Discussion by yw662** [sidorares/node-mysql2#2913] [Announcements]
I know according to the protocol mysql only accept one statement a time.

But is there a way to work around it ?
I tried
```js
for(const q of script.split(';')) {
    await pool.query(q);
}
````

It usually works until I encountered sth. like
```sql
delimiter $$
create procedure foo
begin
select........;
end $$
delimiter ;
````

So, do I really need to find a way to handle this ?

**Comment by sidorares** [comment]
At protocol level it's "only one command at a time" and query command can contain multiple statements - see https://github.com/mysqljs/mysql#multiple-statement-queries

**Comment by yw662** [comment]
but the tricky thing is, that flag does not work when the delimiter change :-)

**Comment by sidorares** [comment]
In your case it's one single statement, and you don't need to set delimiter. See example here - https://github.com/sidorares/node-mysql2/blob/0c422dd9d88a211015ddf4fcc8531048de9b1340/test/integration/connection/test-binary-multiple-results.js#L100-L104

**Comment by yw662** [comment]
But what if I read it from a sql script ?
sth like
```sql
drop table if exists foo
create table foo (
......

drop procedure if exists bar;
delimiter $$
create procedure bar
(......
begin
-- a lot of statements --
end $$
delimiter ;
````

**Comment by sidorares** [comment]
afaik you just use part without delimiter:

```js
conn.query('drop procedure if exists bar');
conn.query(`
create procedure bar
(......
begin
-- a lot of statements --
end
`);
```

**Comment by terrisgit** [comment]
I inherited code that does something like Promise.all( connection.query(), connection.query() )

The SQL statements don't do anything tricky like commit or rollback, so as long as connection.query() "locks" the connection until the result set is loaded into memory from the server, I guess this will work.

The code is written this way because the Promise collection is generated by passing a callback to Array.map() that returns a Promise. Therefore, multiple connection.query()'s run at the same time, more or less.

So far we have not seen any issues but is this supported usage? Is it advantageous any way to

await connection.query(...);
await connection.query(...); ?

**Comment by sidorares** [comment]
>  Is it advantageous any way to
await connection.query(...);
await connection.query(...); ?

no, they'll be executed sequentially anyway. There might be a benefit of doing

```js
await Promise.all([pool.query(...), pool.query(...)])
```
- that way both queries are actually might run in parallel because they are likely sent from different connections

**Comment by terrisgit** [comment]
Thanks! For our multiuser app, we need database connections to be available for incoming requests and therefore can't allow one request to hog connections not to mention the entire mysql server with a swarm of work. Also, for one request, there can only be one database transaction that is rolled back upon failure. Therefore we use synchronous SQL commands for individual API requests. Just giving mysql2 users something to think about when they stumble across this conversation.

**Comment by drwharris** [comment]
I am at a loss. I have tried every which way to make this work but cant.

    DELIMITER $$
    CREATE PROCEDURE `archive_events`()
    BEGIN
        DECLARE max_id INT DEFAULT 0;
        DECLARE EXIT HANDLER FOR SQLEXCEPTION 
        BEGIN
            ROLLBACK;
            DROP TEMPORARY TABLE tmp_ids;
        END;
    
        main:LOOP
            CREATE TEMPORARY TABLE tmp_ids AS SELECT event_id FROM event WHERE event_date < DATE_ADD(NOW(), INTERVAL -151 DAY) ORDER BY event_id LIMIT 1000;
            SELECT MAX(event_id) INTO max_id FROM tmp_ids;
            IF ISNULL(max_id) THEN
                DROP TEMPORARY TABLE tmp_ids;
                LEAVE main;
            END IF;
            
            START TRANSACTION;
                DELETE e FROM event e JOIN tmp_ids t ON e.event_id = t.event_id;
                COMMIT;
            DROP TEMPORARY TABLE tmp_ids;
        END LOOP;
    END$$
    DELIMITER ;
    
If I remove the DELIMITER statements then the ";" cause the connection to think there are multiple statements. If I remove the ";" then the statement isn't valid SQL.

Any suggestions?

**Comment by terrisgit** [comment]
Creating stored procedures with mysql2 is atypical. Have you considered db-migrate? Be sure to omit DELIMITER statements in db-migrate scripts.

This might be helpful: https://stackoverflow.com/questions/47369349/how-does-one-create-a-stored-procedure-in-node-js-using-the-mysql-library

**Comment by yw662** [comment]
If you are executing a script from file and have to stick to the sql script syntax, like what I did, you may need to handle delimiter lines and process the script manually.

It is like, first split the script by ';', then search in each statement for delimiter lines, then replace the delimiter to ';' and remove delimiter line.

**Comment by sidorares** [comment]
@terrisgit what do you mean by atypical? mysql2 is a relatively low level driver and only supports syntax that mysql server understands. I guess most confusion comes from differences that mysql cli adds by itself on the client before sending to the server

**Comment by terrisgit** [comment]
@sidorares I meant that creating a stored procedure as part of a one-time migration is orders of magnitude more common than creating them during normal application activity. That said, yes, you could implement that using mysql2. 

I think @drwharris needs to disable the multipleStatements option when connecting (and of course not using the DELIMITER statements). It's unclear how you are creating the connection.

**Comment by drwharris** [comment]
let statement = -> loaded from file <-
        let dbConnection = {
            host: myHost,
            user: myUsername,
            password: myPassword,
            waitForConnections: true,
            connectionLimit: 1,
            queueLimit: 1,
            database: dbName
        };
        db = await mysql.createConnection(dbConnection);
        await db.query(statement);

I can set the statement to be as follows (removing the delimiter lines) but it then complains about the semi colons (It thinks its multi-statement?). If I remove the semi colons its complains about invalid SQL at the 2nd declare statement.

    CREATE PROCEDURE `archive_events`()
    BEGIN
        DECLARE max_id INT DEFAULT 0;
        DECLARE EXIT HANDLER FOR SQLEXCEPTION 
        BEGIN
            ROLLBACK;
            DROP TEMPORARY TABLE tmp_ids;
        END;
    
        main:LOOP
            CREATE TEMPORARY TABLE tmp_ids AS SELECT event_id FROM event WHERE event_date < DATE_ADD(NOW(), INTERVAL -151 DAY) ORDER BY event_id LIMIT 1000;
            SELECT MAX(event_id) INTO max_id FROM tmp_ids;
            IF ISNULL(max_id) THEN
                DROP TEMPORARY TABLE tmp_ids;
                LEAVE main;
            END IF;
            
            START TRANSACTION;
                DELETE e FROM event e JOIN tmp_ids t ON e.event_id = t.event_id;
                COMMIT;
            DROP TEMPORARY TABLE tmp_ids;
        END LOOP;
    END

**Comment by drwharris** [comment]
Would it not be possible to have a "raw" option to just pass a statement to the DB and let it to its job without client side parsing, etc?

**Comment by yw662** [comment]
> Would it not be possible to have a "raw" option to just pass a statement to the DB and let it to its job without client side parsing, etc?

"without client side parsing"
Actually the delimiter is totally done by client side parsing and mysql server know nothing about it.

**Comment by drwharris** [comment]
OK I think i solved it. You need to terminate each line with a \r\n

**Comment by JaoodxD** [comment]
@drwharris could you please provide minimal example?

**Comment by drwharris** [comment]
> @drwharris could you please provide minimal example?

4y on? One of my comments had exactly what I was trying to do.

**Comment by JaoodxD** [comment]
> > @drwharris could you please provide minimal example?
> 
> 4y on? One of my comments had exactly what I was trying to do.

Yes, this is still an issue.
Could you clarify what you meant by saying 'terminate each line with a \r\n'? 
I thought each line in a text file is already terminated with this sequence by default on Windows and with '\n' on unix-like systems.
